core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -370,14 +410,47 @@ protected void stopServing() throws Exception {
         Thread.currentThread().interrupt();
       }
     }
-    if (mJvmPauseMonitor != null) {
-      mJvmPauseMonitor.stop();
-    }
+  }
+
+  /**
+   * Stops serving, trying stop RPC server and web ui server and letting {@link MetricsSystem} stop
+   * all the sinks.
+   */
+  protected void stopServing() throws Exception {
+    stopLeaderServing();
+    stopCommonServices(null, Collections.EMPTY_LIST);
+  }
+
+  protected void stopServingWebServer() throws Exception {
     if (mWebServer != null) {
       mWebServer.stop();
       mWebServer = null;
     }
-    MetricsSystem.stopSinks();
+  }
+
+  protected void stopJvmMonitorProcess() {
+    if (mJvmPauseMonitor != null) {
+      mJvmPauseMonitor.stop();
+    }
+  }
+
+  protected void stopCommonServices(List<String> includeServiceNames,
+      List<String> excludeServiceNames) throws Exception {
+    for (Map.Entry<String, Pair<CheckedVoidFunction, CheckedVoidFunction>> entry :
+        mCommonServices.entrySet()) {
+      if (includeServiceNames != null && !includeServiceNames.contains(entry.getKey())) {
+        continue;
+      }
+      if (excludeServiceNames.contains(entry.getKey())) {
+        continue;
+      }
+      try {
+        entry.getValue().getSecond().run();
+      } catch (Exception e) {
+        LOG.error("{} stop error", entry.getKey(), e);
+        throw e;
+      }
+    }
   }
```

Comment:
See comment about separating the `METRICS_SINK`, `WEB`, and `JVM_PAUSE_MONITOR` into distinct services. You already separated our both the `WEB` and `JVM_PAUSE_MONITOR` ones here anyway.